File: backend/app/docs/api/examples/data_import_examples.py
"""
Data Import API Examples - Python Client Implementation.

This file provides complete working examples for all data import API endpoints.
"""

import requests
import json
import base64
import csv
import io
from typing import List, Dict, Any, Optional
from datetime import datetime
import time


class DataImportClient:
    """Client for interacting with the Data Import API."""
    
    def __init__(self, base_url: str, token: str, client_id: str, engagement_id: str):
        """
        Initialize the Data Import API client.
        
        Args:
            base_url: API base URL (e.g., "https://api.yourdomain.com")
            token: Bearer authentication token
            client_id: Client account ID
            engagement_id: Engagement ID
        """
        self.base_url = base_url.rstrip('/')
        self.headers = {
            "Authorization": f"Bearer {token}",
            "X-Client-Account-ID": client_id,
            "X-Engagement-ID": engagement_id,
            "Content-Type": "application/json"
        }
    
    def import_csv_file(self, file_path: str, import_type: str) -> Dict[str, Any]:
        """
        Import a CSV file from disk.
        
        Args:
            file_path: Path to the CSV file
            import_type: Type of import (servers, applications, etc.)
            
        Returns:
            API response containing import ID and flow ID
        """
        # Read and parse CSV file
        with open(file_path, 'r') as f:
            reader = csv.DictReader(f)
            data = list(reader)
        
        # Get file metadata
        import os
        file_stats = os.stat(file_path)
        
        # Prepare request
        request_data = {
            "file_data": data,
            "metadata": {
                "filename": os.path.basename(file_path),
                "size": file_stats.st_size,
                "type": "text/csv"
            },
            "upload_context": {
                "intended_type": import_type,
                "upload_timestamp": datetime.utcnow().isoformat() + "Z"
            }
        }
        
        # Send request
        response = requests.post(
            f"{self.base_url}/api/v1/data-import/store-import",
            headers=self.headers,
            json=request_data
        )
        
        response.raise_for_status()
        return response.json()
    
    def import_data_direct(self, data: List[Dict[str, Any]], filename: str, import_type: str) -> Dict[str, Any]:
        """
        Import data directly without reading from file.
        
        Args:
            data: List of dictionaries representing CSV rows
            filename: Filename for tracking
            import_type: Type of import
            
        Returns:
            API response
        """
        request_data = {
            "file_data": data,
            "metadata": {
                "filename": filename,
                "size": len(json.dumps(data).encode('utf-8')),
                "type": "text/csv"
            },
            "upload_context": {
                "intended_type": import_type,
                "upload_timestamp": datetime.utcnow().isoformat() + "Z"
            }
        }
        
        response = requests.post(
            f"{self.base_url}/api/v1/data-import/store-import",
            headers=self.headers,
            json=request_data
        )
        
        response.raise_for_status()
        return response.json()
    
    def get_import_status(self, import_id: str) -> Dict[str, Any]:
        """Get the status of an import operation."""
        response = requests.get(
            f"{self.base_url}/api/v1/data-import/import/{import_id}/status",
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()
    
    def get_latest_import(self) -> Dict[str, Any]:
        """Get the most recent import for the current context."""
        response = requests.get(
            f"{self.base_url}/api/v1/data-import/latest-import",
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()
    
    def wait_for_import_completion(self, import_id: str, timeout: int = 300) -> Dict[str, Any]:
        """
        Wait for an import to complete with polling.
        
        Args:
            import_id: Import ID to monitor
            timeout: Maximum wait time in seconds
            
        Returns:
            Final import status
        """
        start_time = time.time()
        
        while time.time() - start_time < timeout:
            status = self.get_import_status(import_id)
            
            if status["import_status"]["status"] in ["completed", "failed"]:
                return status
            
            # Wait before next poll
            time.sleep(5)
        
        raise TimeoutError(f"Import {import_id} did not complete within {timeout} seconds")
    
    def cancel_import(self, import_id: str) -> Dict[str, Any]:
        """Cancel an ongoing import operation."""
        response = requests.delete(
            f"{self.base_url}/api/v1/data-import/import/{import_id}",
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()
    
    def retry_failed_import(self, import_id: str) -> Dict[str, Any]:
        """Retry a failed import operation."""
        response = requests.post(
            f"{self.base_url}/api/v1/data-import/import/{import_id}/retry",
            headers=self.headers
        )
        response.raise_for_status()
        return response.json()


# Example Usage
def main():
    """Demonstrate usage of the Data Import API client."""
    
    # Initialize client
    client = DataImportClient(
        base_url="https://api.yourdomain.com",
        token="your-bearer-token",
        client_id="1",
        engagement_id="1"
    )
    
    # Example 1: Import server data directly
    print("Example 1: Direct Server Data Import")
    print("-" * 50)
    
    server_data = [
        {
            "server_name": "prod-web-01",
            "ip_address": "10.0.1.10",
            "os": "Ubuntu 20.04",
            "cpu_cores": 8,
            "memory_gb": 16,
            "storage_gb": 500,
            "environment": "production",
            "business_unit": "Sales",
            "application": "Web Portal"
        },
        {
            "server_name": "prod-db-01",
            "ip_address": "10.0.1.20",
            "os": "RHEL 8",
            "cpu_cores": 16,
            "memory_gb": 64,
            "storage_gb": 2000,
            "environment": "production",
            "business_unit": "Sales",
            "application": "Customer Database"
        },
        {
            "server_name": "dev-app-01",
            "ip_address": "10.0.2.10",
            "os": "CentOS 7",
            "cpu_cores": 4,
            "memory_gb": 8,
            "storage_gb": 200,
            "environment": "development",
            "business_unit": "Engineering",
            "application": "API Service"
        }
    ]
    
    try:
        result = client.import_data_direct(
            data=server_data,
            filename="servers_inventory.csv",
            import_type="servers"
        )
        print(f"Import successful!")
        print(f"Import ID: {result['data_import_id']}")
        print(f"Flow ID: {result['flow_id']}")
        print(f"Total Records: {result['total_records']}")
        print()
        
        # Wait for completion
        print("Waiting for import to complete...")
        final_status = client.wait_for_import_completion(result['data_import_id'])
        print(f"Import completed with status: {final_status['import_status']['status']}")
        
    except requests.exceptions.HTTPError as e:
        if e.response.status_code == 409:
            print("Error: An incomplete discovery flow already exists")
            error_detail = e.response.json()
            print(f"Existing flow: {error_detail.get('existing_flow', {}).get('flow_id')}")
            print("Recommendations:")
            for rec in error_detail.get('recommendations', []):
                print(f"  - {rec}")
        else:
            print(f"HTTP Error: {e}")
    
    print()
    
    # Example 2: Import from CSV file
    print("Example 2: Import from CSV File")
    print("-" * 50)
    
    # Create a sample CSV file
    csv_content = """app_name,version,server_name,technology_stack,criticality
CustomerPortal,2.3.4,prod-web-01,Java Spring Boot,High
InventoryAPI,1.8.2,prod-app-01,Python FastAPI,Medium
AnalyticsEngine,3.0.1,prod-analytics-01,Apache Spark,High
MobileBackend,1.5.0,prod-mobile-01,Node.js,Medium
PaymentGateway,2.1.0,prod-payment-01,Java Spring,Critical"""
    
    with open("applications.csv", "w") as f:
        f.write(csv_content)
    
    try:
        result = client.import_csv_file(
            file_path="applications.csv",
            import_type="applications"
        )
        print(f"Applications import successful!")
        print(f"Import ID: {result['data_import_id']}")
        
    except Exception as e:
        print(f"Error importing applications: {e}")
    
    print()
    
    # Example 3: Check import status
    print("Example 3: Check Import Status")
    print("-" * 50)
    
    try:
        latest = client.get_latest_import()
        if latest['success'] and latest['import_metadata']:
            import_id = latest['import_metadata']['import_id']
            status = client.get_import_status(import_id)
            print(f"Import ID: {import_id}")
            print(f"Status: {status['import_status']['status']}")
            print(f"Progress: {status['import_status'].get('progress', 0)}%")
            print(f"Current Phase: {status['import_status'].get('current_phase', 'Unknown')}")
    except Exception as e:
        print(f"Error checking status: {e}")
    
    print()
    
    # Example 4: Error handling for validation errors
    print("Example 4: Handling Validation Errors")
    print("-" * 50)
    
    invalid_data = [
        {
            "server_name": "",  # Invalid: empty server name
            "ip_address": "not-an-ip",  # Invalid: bad IP format
            "os": "Ubuntu 20.04"
        }
    ]
    
    try:
        result = client.import_data_direct(
            data=invalid_data,
            filename="invalid_servers.csv",
            import_type="servers"
        )
    except requests.exceptions.HTTPError as e:
        print(f"Expected validation error: {e.response.status_code}")
        print(f"Error details: {e.response.json()}")
    
    print()
    
    # Example 5: Batch import with progress monitoring
    print("Example 5: Batch Import with Progress Monitoring")
    print("-" * 50)
    
    # Generate larger dataset
    large_dataset = []
    for i in range(100):
        large_dataset.append({
            "server_name": f"server-{i:03d}",
            "ip_address": f"10.0.{i // 256}.{i % 256}",
            "os": "Ubuntu 20.04" if i % 2 == 0 else "RHEL 8",
            "cpu_cores": 4 + (i % 4) * 4,
            "memory_gb": 8 + (i % 8) * 8,
            "storage_gb": 100 + (i % 10) * 100
        })
    
    try:
        result = client.import_data_direct(
            data=large_dataset,
            filename="large_server_inventory.csv",
            import_type="servers"
        )
        
        import_id = result['data_import_id']
        print(f"Large import started: {import_id}")
        
        # Monitor progress
        while True:
            status = client.get_import_status(import_id)
            current_status = status['import_status']['status']
            progress = status['import_status'].get('progress', 0)
            
            print(f"\rProgress: {progress}% - Status: {current_status}", end="")
            
            if current_status in ["completed", "failed"]:
                print()  # New line
                break
            
            time.sleep(2)
        
        print(f"Final status: {current_status}")
        
    except Exception as e:
        print(f"Error in batch import: {e}")


if __name__ == "__main__":
    main()

File: backend/app/docs/api/examples/data_import_examples.ts
/**
 * Data Import API Examples - TypeScript/JavaScript Client Implementation
 * 
 * This file provides complete working examples for all data import API endpoints.
 */

interface FileMetadata {
  filename: string;
  size: number;
  type: string;
}

interface UploadContext {
  intended_type: string;
  validation_flow_id?: string;
  upload_timestamp: string;
}

interface DataImportRequest {
  file_data: Record<string, unknown>[];
  metadata: FileMetadata;
  upload_context: UploadContext;
  client_id?: string;
  engagement_id?: string;
}

interface DataImportResponse {
  success: boolean;
  message: string;
  data_import_id: string;
  flow_id: string;
  total_records: number;
  import_type: string;
  next_steps: string[];
}

interface ImportStatus {
  import_id: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  progress?: number;
  current_phase?: string;
  updated_at: string;
}

class DataImportClient {
  private baseUrl: string;
  private headers: Record<string, string>;

  constructor(
    baseUrl: string,
    token: string,
    clientId: string,
    engagementId: string
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.headers = {
      'Authorization': `Bearer ${token}`,
      'X-Client-Account-ID': clientId,
      'X-Engagement-ID': engagementId,
      'Content-Type': 'application/json'
    };
  }

  /**
   * Import CSV data directly
   */
  async importData(
    data: Record<string, unknown>[],
    filename: string,
    importType: string
  ): Promise<DataImportResponse> {
    const request: DataImportRequest = {
      file_data: data,
      metadata: {
        filename,
        size: JSON.stringify(data).length,
        type: 'text/csv'
      },
      upload_context: {
        intended_type: importType,
        upload_timestamp: new Date().toISOString()
      }
    };

    const response = await fetch(`${this.baseUrl}/api/v1/data-import/store-import`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify(request)
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(`Import failed: ${error.message || response.statusText}`);
    }

    return response.json();
  }

  /**
   * Import CSV file from browser file input
   */
  async importFile(file: File, importType: string): Promise<DataImportResponse> {
    // Parse CSV file
    const text = await file.text();
    const data = this.parseCSV(text);

    const request: DataImportRequest = {
      file_data: data,
      metadata: {
        filename: file.name,
        size: file.size,
        type: file.type || 'text/csv'
      },
      upload_context: {
        intended_type: importType,
        upload_timestamp: new Date().toISOString()
      }
    };

    const response = await fetch(`${this.baseUrl}/api/v1/data-import/store-import`, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify(request)
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(`Import failed: ${error.message || response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get import status
   */
  async getImportStatus(importId: string): Promise<ImportStatus> {
    const response = await fetch(
      `${this.baseUrl}/api/v1/data-import/import/${importId}/status`,
      { headers: this.headers }
    );

    if (!response.ok) {
      throw new Error(`Failed to get status: ${response.statusText}`);
    }

    const result = await response.json();
    return result.import_status;
  }

  /**
   * Get latest import for current context
   */
  async getLatestImport(): Promise<DataImportResponse> {
    const response = await fetch(
      `${this.baseUrl}/api/v1/data-import/latest-import`,
      { headers: this.headers }
    );

    if (!response.ok) {
      throw new Error(`Failed to get latest import: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Wait for import completion with polling
   */
  async waitForCompletion(
    importId: string,
    options: { timeout?: number; pollInterval?: number } = {}
  ): Promise<ImportStatus> {
    const { timeout = 300000, pollInterval = 5000 } = options;
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const status = await this.getImportStatus(importId);
      
      if (status.status === 'completed' || status.status === 'failed') {
        return status;
      }

      await new Promise(resolve => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Import ${importId} did not complete within ${timeout}ms`);
  }

  /**
   * Cancel an import
   */
  async cancelImport(importId: string): Promise<void> {
    const response = await fetch(
      `${this.baseUrl}/api/v1/data-import/import/${importId}`,
      {
        method: 'DELETE',
        headers: this.headers
      }
    );

    if (!response.ok) {
      throw new Error(`Failed to cancel import: ${response.statusText}`);
    }
  }

  /**
   * Retry a failed import
   */
  async retryImport(importId: string): Promise<DataImportResponse> {
    const response = await fetch(
      `${this.baseUrl}/api/v1/data-import/import/${importId}/retry`,
      {
        method: 'POST',
        headers: this.headers
      }
    );

    if (!response.ok) {
      throw new Error(`Failed to retry import: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Parse CSV text to array of objects
   */
  private parseCSV(text: string): Record<string, unknown>[] {
    const lines = text.split('\n').filter(line => line.trim());
    if (lines.length === 0) return [];

    const headers = lines[0].split(',').map(h => h.trim());
    const data: Record<string, unknown>[] = [];

    for (let i = 1; i < lines.length; i++) {
      const values = lines[i].split(',').map(v => v.trim());
      const row: Record<string, unknown> = {};
      
      headers.forEach((header, index) => {
        row[header] = values[index] || '';
      });
      
      data.push(row);
    }

    return data;
  }
}

// Example Usage
async function demonstrateDataImportAPI() {
  // Initialize client
  const client = new DataImportClient(
    'https://api.yourdomain.com',
    'your-bearer-token',
    '1',
    '1'
  );

  console.log('=== Data Import API Examples ===\n');

  // Example 1: Direct data import
  console.log('Example 1: Direct Server Data Import');
  console.log('-'.repeat(50));

  const serverData = [
    {
      server_name: 'prod-web-01',
      ip_address: '10.0.1.10',
      os: 'Ubuntu 20.04',
      cpu_cores: 8,
      memory_gb: 16,
      storage_gb: 500,
      environment: 'production',
      business_unit: 'Sales',
      application: 'Web Portal'
    },
    {
      server_name: 'prod-db-01',
      ip_address: '10.0.1.20',
      os: 'RHEL 8',
      cpu_cores: 16,
      memory_gb: 64,
      storage_gb: 2000,
      environment: 'production',
      business_unit: 'Sales',
      application: 'Customer Database'
    }
  ];

  try {
    const result = await client.importData(
      serverData,
      'servers_inventory.csv',
      'servers'
    );
    
    console.log('Import successful!');
    console.log(`Import ID: ${result.data_import_id}`);
    console.log(`Flow ID: ${result.flow_id}`);
    console.log(`Total Records: ${result.total_records}`);
    console.log('Next Steps:', result.next_steps);

    // Wait for completion
    console.log('\nWaiting for import to complete...');
    const finalStatus = await client.waitForCompletion(result.data_import_id);
    console.log(`Import completed with status: ${finalStatus.status}`);

  } catch (error: unknown) {
    console.error('Import failed:', error.message);
  }

  console.log();

  // Example 2: File upload from browser
  console.log('Example 2: Browser File Upload');
  console.log('-'.repeat(50));

  // Simulate file input
  const csvContent = `app_name,version,server_name,technology_stack,criticality
CustomerPortal,2.3.4,prod-web-01,Java Spring Boot,High
InventoryAPI,1.8.2,prod-app-01,Python FastAPI,Medium
AnalyticsEngine,3.0.1,prod-analytics-01,Apache Spark,High`;

  const file = new File([csvContent], 'applications.csv', { type: 'text/csv' });

  try {
    const result = await client.importFile(file, 'applications');
    console.log('Applications import successful!');
    console.log(`Import ID: ${result.data_import_id}`);
  } catch (error: unknown) {
    console.error('File import failed:', error.message);
  }

  console.log();

  // Example 3: React component example
  console.log('Example 3: React Component Integration');
  console.log('-'.repeat(50));
  console.log(`
import React, { useState } from 'react';
import { DataImportClient } from './DataImportClient';

function DataImportComponent() {
  const [importing, setImporting] = useState(false);
  const [progress, setProgress] = useState(0);
  const [error, setError] = useState<string | null>(null);

  const client = new DataImportClient(
    process.env.REACT_APP_API_URL!,
    localStorage.getItem('authToken')!,
    localStorage.getItem('clientId')!,
    localStorage.getItem('engagementId')!
  );

  const handleFileUpload = async (event: React.ChangeEvent<HTMLInputElement>) => {
    const file = event.target.files?.[0];
    if (!file) return;

    setImporting(true);
    setError(null);

    try {
      // Import the file
      const result = await client.importFile(file, 'servers');
      
      // Monitor progress
      const pollProgress = setInterval(async () => {
        try {
          const status = await client.getImportStatus(result.data_import_id);
          setProgress(status.progress || 0);
          
          if (status.status === 'completed' || status.status === 'failed') {
            clearInterval(pollProgress);
            setImporting(false);
            
            if (status.status === 'failed') {
              setError('Import failed');
            }
          }
        } catch (err) {
          clearInterval(pollProgress);
          setImporting(false);
          setError('Failed to check status');
        }
      }, 2000);

    } catch (err: any) {
      setImporting(false);
      setError(err.message);
    }
  };

  return (
    <div>
      <input
        type="file"
        accept=".csv"
        onChange={handleFileUpload}
        disabled={importing}
      />
      
      {importing && (
        <div>
          <progress value={progress} max={100} />
          <span>{progress}%</span>
        </div>
      )}
      
      {error && <div className="error">{error}</div>}
    </div>
  );
}
`);

  console.log();

  // Example 4: Error handling
  console.log('Example 4: Comprehensive Error Handling');
  console.log('-'.repeat(50));

  try {
    // Attempt import with invalid data
    await client.importData(
      [{ invalid: 'data' }],
      'invalid.csv',
      'servers'
    );
  } catch (error: unknown) {
    if (error.message.includes('incomplete_discovery_flow_exists')) {
      console.log('Handling existing flow conflict:');
      console.log('1. Show user the existing flow');
      console.log('2. Offer to cancel or complete it');
      console.log('3. Retry after resolution');
    } else if (error.message.includes('validation_error')) {
      console.log('Handling validation error:');
      console.log('1. Parse error details');
      console.log('2. Highlight problematic fields');
      console.log('3. Guide user to fix issues');
    } else {
      console.log('Generic error handling:', error.message);
    }
  }

  console.log();

  // Example 5: Batch operations with progress
  console.log('Example 5: Batch Import with Real-time Progress');
  console.log('-'.repeat(50));

  // Generate large dataset
  const largeDataset = Array.from({ length: 1000 }, (_, i) => ({
    server_name: `server-${String(i).padStart(3, '0')}`,
    ip_address: `10.0.${Math.floor(i / 256)}.${i % 256}`,
    os: i % 2 === 0 ? 'Ubuntu 20.04' : 'RHEL 8',
    cpu_cores: 4 + (i % 4) * 4,
    memory_gb: 8 + (i % 8) * 8,
    storage_gb: 100 + (i % 10) * 100
  }));

  try {
    console.log(`Importing ${largeDataset.length} servers...`);
    const result = await client.importData(
      largeDataset,
      'large_inventory.csv',
      'servers'
    );

    // Create progress bar
    const progressBar = (percent: number) => {
      const filled = Math.floor(percent / 2);
      const empty = 50 - filled;
      return `[${'='.repeat(filled)}${' '.repeat(empty)}] ${percent}%`;
    };

    // Monitor with visual progress
    let lastProgress = 0;
    while (true) {
      const status = await client.getImportStatus(result.data_import_id);
      const progress = status.progress || 0;
      
      if (progress !== lastProgress) {
        process.stdout.write(`\r${progressBar(progress)} - ${status.current_phase || 'Processing'}`);
        lastProgress = progress;
      }

      if (status.status === 'completed' || status.status === 'failed') {
        console.log('\nImport completed!');
        break;
      }

      await new Promise(resolve => setTimeout(resolve, 1000));
    }

  } catch (error: unknown) {
    console.error('Batch import failed:', error.message);
  }
}

// Run examples
if (typeof window === 'undefined') {
  // Node.js environment
  demonstrateDataImportAPI().catch(console.error);
}

// Export for use in other modules
export { DataImportClient, demonstrateDataImportAPI };